### PR TITLE
refactor: split `ge_parse` into explicit variants (compressed, uncompressed, uncompressed+hybrid)

### PR DESCRIPTION
**Pull request by theStack**
This PR is a follow-up to #1915, see the suggestion https://github.com/bitcoin-core/secp256k1/pull/1915#pullrequestreview-4962713401.

Splits the `ge_parse` function into three explicit variants for parsing from:
    - compressed public keys (33-bytes, prefixes 0x02 or 0x03)
    - uncompressed public keys (65-bytes, prefix 0x04)
    - uncompressed and hybrid public keys (65-bytes, prefixes 0x04, 0x06 or 0x07)

The first two are the counter-parts for the existing serialization functions `ge_serialize{33,65}`. The latter one is only needed for legacy reasons and likely not needed in the future. Having it named explicitly helps to avoid accepting hybrid pubkeys unintentionally. `ge_parse65` is currently unused outside of tests, but one potential use-case would be the serialization of the prevouts_summary in the silentpayments light-client API PR #1912 (see commit https://github.com/bitcoin-core/secp256k1/pull/1912/changes/82b95fde6fe2c75ec940ade339013939b8ecc5ea), unless we want to define a custom serialization there (where we e.g. remove the constant prefix byte).

### CODE REVIEW

File: src/secp256k1.c [hunk 1]
```diff
@@ -267,12 +267,18 @@ static void secp256k1_pubkey_save(secp256k1_pubkey* pubkey, secp256k1_ge* ge) {
 
 int secp256k1_ec_pubkey_parse(const secp256k1_context* ctx, secp256k1_pubkey* pubkey, const unsigned char *input, size_t inputlen) {
     secp256k1_ge Q;
+    int res;
```
Comment (fjahr) [inline src/secp256k1.c]: super nit: I think it's usually `ret`?

File: src/secp256k1.c [hunk 2]
```diff
@@ -267,12 +267,18 @@ static void secp256k1_pubkey_save(secp256k1_pubkey* pubkey, secp256k1_ge* ge) {
...
 
     VERIFY_CHECK(ctx != NULL);
     ARG_CHECK(pubkey != NULL);
     memset(pubkey, 0, sizeof(*pubkey));
     ARG_CHECK(input != NULL);
-    if (!secp256k1_ge_parse(&Q, input, inputlen)) {
+    switch (inputlen) {
+        case 33: res = secp256k1_ge_parse33(&Q, input); break;
```
Comment (fjahr) [inline src/secp256k1.c]: I like this concise style, just wondering why it's never been used anywhere outside of test code. I guess there is no particular reason?

File: src/tests.c [hunk 3]
```diff
@@ -6840,6 +6840,7 @@ static void test_random_pubkeys(void) {
     secp256k1_ge elem;
     secp256k1_ge elem2;
     unsigned char in[65];
+    int res;
     /* Generate some randomly sized pubkeys. */
     size_t len = testrand_bits(2) == 0 ? 65 : 33;
     if (testrand_bits(2) == 0) {
```
Comment (fjahr) [inline src/tests.c]: With the switch statement added further below I think this randomness here doesn't make sense anymore. If `len` isn't 33 or 65 we just end up hitting the default case there and we are not even calling any library function in the test then.

**Review by fjahr — COMMENTED** [review]
Concept ACK

File: src/secp256k1.c (hunk 1 above)
Comment (theStack) [reply src/secp256k1.c]: Seems that in API functions we use `ret` almost exclusively for the to-be-returned value (i.e. paired with a `return ret;` at the bottom), which doesn't apply here. `res` is not a great name either, decided to be explicit and name it `is_pubkey_valid` (inspired by existing `is_sec_valid` instances in the codebase).

File: src/secp256k1.c (hunk 2 above)
Comment (theStack) [reply src/secp256k1.c]: Fair question, maybe there was no good use-case for it so far, but not sure. Seems there is one switch/case statement in `secp256k1_pippenger_bucket_window_inv`, so there is at least no (unwritten) rule to completely avoid it in (potential) production code. No strong opinion, also happy to change to if/else if other reviewers feel strongly.

In the test I've changed to using a ternary expression, since the default branch is not needed anymore after your suggestion below to limit to 33/65 byte-sized pubkey serialization, so the switch/case became a bit less compelling.

File: src/tests.c (hunk 3 above)
Comment (theStack) [reply src/tests.c]: Good point, pretty pointless now to generate differently-sized serializations and not do anything with them indeed. Adapted and simplified the test accordingly.

**Comment by theStack** [comment]
Thanks for the detailed review @fjahr, addressed your suggestion of simplifying/adapting the test and renamed the res variable in the touched pubkey parsing API function.

File: src/group.h [hunk 4]
```diff
@@ -196,8 +196,14 @@ static void secp256k1_ge_to_bytes_ext(unsigned char *data, const secp256k1_ge *g
...
 
-/** Parse a group element from a 33-byte compressed or 65-byte uncompressed public key. */
-static int secp256k1_ge_parse(secp256k1_ge *elem, const unsigned char *pub, size_t size);
+/** Parse a group element from a 33-byte compressed public key (prefix byte is 0x02 or 0x03). */
+static int secp256k1_ge_parse33(secp256k1_ge *elem, const unsigned char *pub);
+
+/** Parse a group element from a 65-byte uncompressed public key (prefix byte is 0x04). */
+static int secp256k1_ge_parse65(secp256k1_ge *elem, const unsigned char *pub);
+
+/** Parse a group element from a 65-byte uncompressed or hybrid public key (prefix byte is 0x04, 0x06 or 0x07). */
```
Comment (fjahr) [inline src/group.h:205]: nit: Could have mentioned here in the comment as well what's in the commit description, that this still exists for legacy reasons and further usage is discouraged. But feel free to ignore unless you have to retouch.

**Review by fjahr — COMMENTED** [review]
ACK 42dd54e46e4223660fa0f6dedb0d4e3f6b0ea152

File: src/group.h (hunk 4 above)
Comment (theStack) [reply src/group.h]: Good idea, will tackle that on the next push (I expect #1923 to go in first, so I'll have to rebase anyways).

**Comment by theStack** [comment]
Rebased on master and addressed https://github.com/bitcoin-core/secp256k1/pull/1918#discussion_r3819523047.

File: src/group.h [hunk 5]
```diff
@@ -205,8 +205,15 @@ static void secp256k1_ge_to_bytes_ext(unsigned char *data, const secp256k1_ge *g
...
-/** Parse a group element from a 33-byte compressed or 65-byte uncompressed public key. */
-static int secp256k1_ge_parse(secp256k1_ge *elem, const unsigned char *pub, size_t size);
+/** Parse a group element from a 33-byte compressed public key (prefix byte is 0x02 or 0x03). */
+static int secp256k1_ge_parse33(secp256k1_ge *elem, const unsigned char *pub);
+
+/** Parse a group element from a 65-byte uncompressed public key (prefix byte is 0x04). */
+static int secp256k1_ge_parse65(secp256k1_ge *elem, const unsigned char *pub);
+
+/** Parse a group element from a 65-byte uncompressed or hybrid public key (prefix byte is 0x04, 0x06 or 0x07).
+ *  Only intended for legacy public APIs that must accept hybrid keys; new code should use secp256k1_ge_parse65. */
```
Comment (real-or-random) [inline src/group.h]:
```suggestion
 *  Only intended for legacy public APIs that must accept hybrid keys. */
```
New code should use ge_parse66, ge_parse33, or xonly even. It depends on the use case, and I don't think we should comment on that here.

File: src/tests.c [hunk 6]
```diff
@@ -6852,11 +6852,9 @@ static void test_random_pubkeys(void) {
...
     secp256k1_ge elem2;
     unsigned char in[65];
-    /* Generate some randomly sized pubkeys. */
-    size_t len = testrand_bits(2) == 0 ? 65 : 33;
-    if (testrand_bits(2) == 0) {
-        len = testrand_bits(6);
-    }
+    int res;
+    /* Generate some random pubkeys with the two supported serialization sizes. */
+    size_t len = testrand_bits(1) == 0 ? 65 : 33;
```
Comment (real-or-random) [inline src/tests.c:6904]: IIUC `run_ec_pubkey_parse_test` currently checks 64 and 66 as edge cases but not 32 and 34. I think these should be added.

**Review by real-or-random — COMMENTED** [review]
ACK mod nits

File: src/group.h (hunk 5 above)
Comment (theStack) [reply src/group.h]: Makes sense yeah, removed that part.

File: src/tests.c (hunk 6 above)
Comment (theStack) [reply src/tests.c:6904]: Added a commit on top that adds coverage for compressed pubkey parsing/serialization, including the 32 and 34 bytes edge cases, in a similar style as how it's already done for uncompressed pubkeys. It could be even more thorough I guess (e.g. by also testing with the 0x03 prefix, not only 0x02), maybe something for a follow-up.

Comment (fjahr) [reply src/tests.c:6904]: nit: Feels like a helper could have deduplicated these edge case tests quite a bit, but not a blocker for me. Basically this (untested):

```
static void ec_pubkey_parse_expect_fail(const unsigned char *input, size_t inputlen) {
    secp256k1_pubkey pubkey;
    secp256k1_ge ge;
    memset(&pubkey, 0xfe, sizeof(pubkey));
    SECP256K1_CHECKMEM_UNDEFINE(&pubkey, sizeof(pubkey));
    CHECK(secp256k1_ec_pubkey_parse(CTX, &pubkey, input, inputlen) == 0);
    SECP256K1_CHECKMEM_CHECK(&pubkey, sizeof(pubkey));
    CHECK_ILLEGAL(CTX, secp256k1_pubkey_load(CTX, &ge, &pubkey));
}
```

**Comment by fjahr** [comment]
Code review ACK de02108c51ca052245c7ba154b02a890e6815fdb

**Review by real-or-random — APPROVED** [review]
utACK de02108c51ca052245c7ba154b02a890e6815fdb
